MediaBrowser.Common/Net/IPObject.cs

```diff
@@ -0,0 +1,395 @@
+        /// IPv6 Loopback address.
+        /// </summary>
+        protected static readonly byte[] Ipv6Loopback = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1 };
+
+        /// <summary>
+        /// IPv4 Loopback address.
+        /// </summary>
+        protected static readonly byte[] Ipv4Loopback = { 127, 0, 0, 1 };
+
+        /// <summary>
+        /// The network address of this object.
+        /// </summary>
+        private IPObject? _networkAddress;
+
+        /// <summary>
+        /// Gets or sets a user defined value that is associated with this object.
+        /// </summary>
+        public int Tag { get; set; }
+
+        /// <summary>
+        /// Gets or sets the object's IP address.
+        /// </summary>
+        public abstract IPAddress Address { get; set; }
+
+        /// <summary>
+        /// Gets the object's network address.
+        /// </summary>
+        public IPObject NetworkAddress => _networkAddress ??= CalculateNetworkAddress();
+
+        /// <summary>
+        /// Gets or sets the object's IP address.
+        /// </summary>
+        public abstract byte PrefixLength { get; set; }
+
+        /// <summary>
+        /// Gets the AddressFamily of this object.
+        /// </summary>
+        public AddressFamily AddressFamily
+        {
+            get
+            {
+                // Keep terms separate as Address performs other functions in inherited objects.
+                IPAddress address = Address;
+                return address.Equals(IPAddress.None) ? AddressFamily.Unspecified : address.AddressFamily;
+            }
+        }
+
+        /// <summary>
+        /// Returns the network address of an object.
+        /// </summary>
+        /// <param name="address">IP Address to convert.</param>
+        /// <param name="prefixLength">Subnet prefix.</param>
+        /// <returns>IPAddress.</returns>
+        public static (IPAddress Address, byte PrefixLength) NetworkAddressOf(IPAddress address, byte prefixLength)
+        {
+            if (address == null)
+            {
+                throw new ArgumentNullException(nameof(address));
+            }
+
+            if (address.IsIPv4MappedToIPv6)
+            {
+                address = address.MapToIPv4();
+            }
+
+            if (IsLoopback(address))
+            {
+                return (Address: address, PrefixLength: prefixLength);
+            }
+
+            // An ip address is just a list of bytes, each one representing a segment on the network.
+            // This separates the IP address into octets and calculates how many octets will need to be altered or set to zero dependant upon the
+            // prefix length value. eg. /16 on a 4 octet ip4 address (192.168.2.240) will result in the 2 and the 240 being zeroed out.
+            // Where there is not an exact boundary (eg /23), mod is used to calculate how many bits of this value are to be kept.
+
+            byte[] addressBytes = address.GetAddressBytes();
```

Comment:
Maybe use a stackalloc here

Reply:
Don't understand how that would be implemented. 

GetAddressBytes() is a method of IPAddress that allocates the memory internally and returns the value)

Reply:
https://docs.microsoft.com/en-us/dotnet/api/system.net.ipaddress.trywritebytes?view=net-5.0#System_Net_IPAddress_TryWriteBytes_System_Span_System_Byte__System_Int32__

https://gitlab.com/Bond_009/T1B9.Utils/-/blob/master/src/T1B9.Utils/IPAddressExtensions.cs#L49-54